src/libraries/System.IO.FileSystem/src/System/IO/FileStatus.Unix.cs

```diff
@@ -9,227 +9,316 @@ internal struct FileStatus
+            // So we need to try to refresh both caches
+            EnsureCachesInitialized(path, continueOnError);
+            return _isDirectory; // Value should be set in Refresh
         }
 
-        internal DateTimeOffset GetLastAccessTime(ReadOnlySpan<char> path, bool continueOnError = false)
+        internal bool IsHidden(ReadOnlySpan<char> path, bool continueOnError = false)
         {
-            EnsureStatInitialized(path, continueOnError);
-            if (!_exists)
-                return DateTimeOffset.FromFileTime(0);
-            return UnixTimeToDateTimeOffset(_fileStatus.ATime, _fileStatus.ATimeNsec);
+            // We only check for the hidden flag in the main path without following symbolic links
+            if (!IsMainCacheValid)
+            {
+                if (!TryRefreshMainCache(path))
+                {
+                    if (!continueOnError)
+                    {
+                        ThrowOnCacheInitializationError(path);
+                    }
+                }
+            }
+            return HasHiddenFlag;
         }
 
-        internal void SetLastAccessTime(string path, DateTimeOffset time) => SetAccessOrWriteTime(path, time, isAccessTime: true);
+        internal bool IsReadOnly(ReadOnlySpan<char> path, bool continueOnError = false)
+        {
+            // We only check for the readonly flag in the main path without following symbolic links
+            if (!IsMainCacheValid)
+            {
+                if (!TryRefreshMainCache(path))
+                {
+                    if (!continueOnError)
+                    {
+                        ThrowOnCacheInitializationError(path);
+                    }
+                }
+            }
+            return HasReadOnlyFlag;
+        }
 
-        internal DateTimeOffset GetLastWriteTime(ReadOnlySpan<char> path, bool continueOnError = false)
+        internal bool IsSymbolicLink(ReadOnlySpan<char> path, bool continueOnError = false)
         {
-            EnsureStatInitialized(path, continueOnError);
-            if (!_exists)
-                return DateTimeOffset.FromFileTime(0);
-            return UnixTimeToDateTimeOffset(_fileStatus.MTime, _fileStatus.MTimeNsec);
+            // We only check for the symbolic link flag in the main path without following symbolic links
+            if (!IsMainCacheValid)
+            {
+                if (!TryRefreshMainCache(path))
+                {
+                    if (!continueOnError)
+                    {
+                        ThrowOnCacheInitializationError(path);
+                    }
+                }
+            }
+            return HasSymbolicLinkFlag;
         }
 
-        internal void SetLastWriteTime(string path, DateTimeOffset time) => SetAccessOrWriteTime(path, time, isAccessTime: false);
+        internal void RefreshCaches(ReadOnlySpan<char> path)
```

Comment:
as mentioned before, it would be great if we could move all the methods that need to call this particular method inside `FileStatus` and make it private, so the cache would become an implementation detail to the consumers

Reply:
There is a difference between called `EnsureCachesInitialized` and `RefreshCaches`: the former will check if they are initialized, and if not, it will refresh them. The second one will _always_ refresh them.
Maybe I can pass a boolean to indicate if we want to force refresh or not. That way, only one method is visible outside.